apps/omg_watcher/lib/exit_processor/core.ex

```diff
@@ -83,8 +113,64 @@ defmodule OMG.Watcher.ExitProcessor.Core do
     {%{state | exits: Map.merge(exits, new_exits_map)}, db_updates}
   end
 
-  defp parse_contract_status({@zero_address, _contract_token, _contract_amount}), do: false
-  defp parse_contract_status({_contract_owner, _contract_token, _contract_amount}), do: true
+  defp parse_contract_exit_status({@zero_address, _contract_token, _contract_amount}), do: false
+  defp parse_contract_exit_status({_contract_owner, _contract_token, _contract_amount}), do: true
+
+  # TODO: syncing problem (look new exits)
+  @doc """
+   Add new in flight exits from Ethereum events into tracked state.
+  """
+  @spec new_in_flight_exits(t(), list(map()), list(map())) :: {t(), list()} | {:error, :unexpected_events}
+  def new_in_flight_exits(state, new_ifes_events, contract_statuses)
+
+  def new_in_flight_exits(_state, new_ifes_events, contract_statuses)
+      when length(new_ifes_events) != length(contract_statuses),
+      do: {:error, :unexpected_events}
+
+  def new_in_flight_exits(%__MODULE__{in_flight_exits: ifes} = state, new_ifes_events, contract_statuses) do
+    new_ifes_kv_pairs =
+      new_ifes_events
+      |> Enum.zip(contract_statuses)
+      |> Enum.map(fn {%{tx_bytes: tx_bytes, signatures: signatures}, {timestamp, contract_ife_id} = contract_status} ->
+        is_active = parse_contract_in_flight_exit_status(contract_status)
+        InFlightExitInfo.new(tx_bytes, signatures, contract_ife_id, timestamp, is_active)
+      end)
+
+    db_updates =
+      new_ifes_kv_pairs
+      |> Enum.map(&InFlightExitInfo.make_db_update/1)
+
+    new_ifes = new_ifes_kv_pairs |> Map.new()
+
+    {%{state | in_flight_exits: Map.merge(ifes, new_ifes)}, db_updates}
+  end
+
+  defp parse_contract_in_flight_exit_status({timestamp, _contract_id}), do: timestamp != 0
+
+  @doc """
+    Add piggybacks from Ethereum events into tracked state.
+  """
+  @spec new_piggybacks(t(), [%{tx_hash: tx_hash(), output_index: output_offset()}]) :: {t(), list()}
+  def new_piggybacks(%__MODULE__{} = state, piggybacks) do
+    {updated_state, updated_pairs} = piggybacks |> Enum.reduce({state, %{}}, &process_piggyback/2)
+
+    {updated_state, Enum.map(updated_pairs, &InFlightExitInfo.make_db_update/1)}
+  end
+
+  defp process_piggyback(
+         %{tx_hash: tx_hash, output_index: output_index},
+         {%__MODULE__{in_flight_exits: ifes} = state, db_updates}
+       ) do
+    with {:ok, ife} <- Map.fetch(ifes, tx_hash),
+         {:ok, updated_ife} <- InFlightExitInfo.piggyback(ife, output_index) do
+      updated_state = %{state | in_flight_exits: Map.put(ifes, tx_hash, updated_ife)}
+      {updated_state, Map.put(db_updates, tx_hash, updated_ife)}
+    else
+      # if impossible to piggyback - do nothing
```

Comment:
shouldn't this be treated as an unrecoverable error, i.e. we have `=` pattern matches instead of a `with`?

Reply:
Yeah, I believe this one is non-recoverable. I can imagine a situation where we are processing IFE and PB in wrong order.